metricflow/query/query_parser.py

```diff
@@ -643,59 +679,90 @@ def _parse_metric_names(
+                    )
+                structured_name = StructuredLinkableSpecName(
+                    entity_link_names=parsed_name.entity_link_names,
+                    element_name=parsed_name.element_name,
+                    time_granularity=linkable_element.grain,
+                    date_part=linkable_element.date_part,
+                )
+                structured_names.append(structured_name)
 
         dimension_specs = []
         time_dimension_specs = []
         partial_time_dimension_specs = []
         entity_specs = []
 
-        for qualified_name in qualified_linkable_names:
-            structured_name = StructuredLinkableSpecName.from_name(qualified_name)
+        for structured_name in structured_names:
             element_name = structured_name.element_name
             entity_links = tuple(EntityReference(element_name=x) for x in structured_name.entity_link_names)
             # Create the spec based on the type of element referenced.
             if TimeDimensionReference(element_name=element_name) in self._known_time_dimension_element_references:
-                if structured_name.time_granularity:
+                if structured_name.time_granularity and not structured_name.date_part:
                     time_dimension_specs.append(
                         TimeDimensionSpec(
                             element_name=element_name,
                             entity_links=entity_links,
                             time_granularity=structured_name.time_granularity,
                         )
                     )
+                # If date part is passed, remove requested granularity (to be overridden with default).
                 else:
                     partial_time_dimension_specs.append(
                         PartialTimeDimensionSpec(
                             element_name=element_name,
                             entity_links=entity_links,
+                            date_part=structured_name.date_part,
                         )
                     )
             elif DimensionReference(element_name=element_name) in self._known_dimension_element_references:
                 dimension_specs.append(DimensionSpec(element_name=element_name, entity_links=entity_links))
             elif EntityReference(element_name=element_name) in self._known_entity_element_references:
                 entity_specs.append(EntitySpec(element_name=element_name, entity_links=entity_links))
             else:
+                valid_group_bys_for_metrics = self._metric_lookup.element_specs_for_metrics(list(metric_references))
                 valid_group_by_names_for_metrics = sorted(
-                    x.qualified_name for x in self._metric_lookup.element_specs_for_metrics(list(metric_references))
+                    list(
+                        set(
+                            x.qualified_name if qualified_linkable_names else x.element_name
```

Comment:
I'm really confused by this. It looks like these are different names. The qualified name includes the spec path but the element name does not, right?

Reply:
Correct!
If `qualified_linkable_names` is passed, we're dealing with the `group_by_names` query param - a list of string `qualified_names`. In that case, we want to check those against the qualified names allowed for the metric.
Otherwise, we know `linkable_elements` was passed into this function, which means the user used the `group_by` query param. That means we have objects instead of strings, and the object name is the `element_name`, not the `qualified_name`. So we want to check those names against the allowed `element_names` for the metric.
Also used a `set` here since we might have repeated `element_names` in the same query.